### PR TITLE
Add RFC 024 - Extension Loading

### PR DESCRIPTION
**Pull request by MMeent**
## Describe your changes

The RFC for our version of custom extension loading, as discussed on the offsite in April.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by hlinnaka — COMMENTED** [review]
Looks good as far as this goes. Lots of open questions remain:

* How is the list of available extensions managed? How does the user add their own extension to the Extension Store?
* How are the extensions packaged? Zip files? Docker images?
* What is the Extension Store? An S3 bucket? A completely new server application that we build?
* How do we hook into extension loading, so that we can download the extension when it's used?

File: docs/rfcs/024-extension-loading.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+time.
+
+That means, we first download the extensions required to start the PostMaster
+(`shared_preload_libraries` and their dependencies), then the libraries required
+before a backend can start processing user input (`preload_libraries` and
+dependencies), and then (with network limits applied) the remainder of the
+configured extensions, with prioritization for installed extensions.
+
+If PostgreSQL tries to load a library that is not yet fully on disk, it will
+ask `compute_ctl` first if the extension has been downloaded yet, and will wait
```
Comment (vadim2404) [inline docs/rfcs/024-extension-loading.md:59]: Where will we store such libraries? do we need to have some kind of auth for this component to be able to download an extension that belongs to this customer?

File: docs/rfcs/024-extension-loading.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+      duplicate extensions.
+    - **UNCHECKED:** Compute instance migration may not work nicely with
+      networked IOs
+
+
+### Idea extensions
+
+The extension store does not have to be S3 directly, but could be a Node-local
+caching service on top of S3. This would reduce the load on the network for
+popular extensions.
```
Comment (lubennikovaav) [inline docs/rfcs/024-extension-loading.md:209]: Wouldn't it be easier to always pre-download all extensions to some node-local cache?
- when we spin a new node - download all extension files to node cache;
- when user uploads new extension to s3, push it to all nodes caches immediately.

Pros:
- `compute_ctl` won't have to care about access to s3
Cons:
- extra space on nodes required.
Though, typical size of extension is not too large.

**Comment by lubennikovaav** [comment]
I've added a chapter about "extension store` organization.
@hlinnaka , I've tried to address there all these [questions](https://github.com/neondatabase/neon/pull/4158#pullrequestreview-1414310007) 


also, force pushed the branch to pass updated CI

File: docs/rfcs/024-extension-loading.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+If PostgreSQL tries to load a library that is not yet fully on disk, it will
+ask `compute_ctl` first if the extension has been downloaded yet, and will wait
+for `compute_ctl` to finish downloading that extension. `compute_ctl` will
+prioritize downloading that extension over other extensions that were not yet
+requested.
+
+#### Workflow
+
+```mermaid
+sequenceDiagram
```
Comment (vadim2404) [inline docs/rfcs/024-extension-loading.md:67]: Does it mean that we will pre-download all users' extensions before starting compute? 

What will happen if the user executes the `CREATE EXTENSION` query? Will we download sth or will rely on pre-downloaded info?

Comment (lubennikovaav) [reply docs/rfcs/024-extension-loading.md:67]: >Does it mean that we will pre-download all users' extensions before starting compute?

We pre-download all control files to know which extensions are available.

> What will happen if the user executes the CREATE EXTENSION query? Will we download sth or will rely on pre-downloaded info?

If control file is available (which means that this is a legit extension), but its sql file and/or library is not found, 
user will request download from `compute_ctl` and wait till the files are downloaded and unpacked into their paths.

Comment (vadim2404) [reply docs/rfcs/024-extension-loading.md:67]: understood, thanks!

**Pull request closed without merging**